features/nvstore/source/nvstore.cpp

```diff
@@ -0,0 +1,993 @@
+    }
+
+    _lock.shared_unlock();
+    return ret;
+}
+
+// Start of API functions
+
+int NVStore::get(uint16_t key, uint16_t buf_len_bytes, uint32_t *buf, uint16_t &actual_len_bytes)
+{
+    return do_get(key, buf_len_bytes, buf, actual_len_bytes, 0);
+}
+
+int NVStore::get_item_size(uint16_t key, uint16_t &actual_len_bytes)
+{
+    return do_get(key, 0, NULL, actual_len_bytes, 1);
+}
+
+int NVStore::do_set(uint16_t key, uint16_t buf_len_bytes, const uint32_t *buf, uint16_t flags)
+{
+    int ret = NVSTORE_SUCCESS;
+    uint32_t record_offset, record_size, new_free_space;
+    uint32_t next_offset;
+    uint8_t save_active_area;
+
+    if (!_init_done) {
+        ret = init();
+        if (ret != NVSTORE_SUCCESS) {
+            return ret;
+        }
+    }
+
+    if (key >= _max_keys) {
+        return NVSTORE_BAD_VALUE;
+    }
+
+    if (!buf)
+        buf_len_bytes = 0;
+
+    if (buf_len_bytes && !is_buf_aligned(buf, sizeof(uint32_t))) {
+        return NVSTORE_BUFF_NOT_ALIGNED;
+    }
+
+    if ((flags & DELETE_ITEM_FLAG) && !_offset_by_key[key]) {
+        return NVSTORE_NOT_FOUND;
+    }
+
+    if (_offset_by_key[key] & OFFS_BY_KEY_SET_ONCE_MASK) {
+        return NVSTORE_ALREADY_EXISTS;
+    }
+
+    // writers do not lock each other exclusively, but can operate in parallel.
+    // Shared lock is in order to prevent GC from operating (which uses exclusive lock).
+    _lock.shared_lock();
+
+    save_active_area = _active_area;
+    record_size = align_up(sizeof(record_header_t) + buf_len_bytes, FLASH_MINIMAL_PROG_UNIT);
+
+    // Parallel operation of writers is allowed due to this atomic operation. This operation
+    // produces an offset on which each writer can work separately, without being interrupted
+    // by the other writer. The only mutual resource here is _free_space_offset - which
+    // gets the correct value because of this atomic increment.
+    new_free_space = safe_increment(_free_space_offset, record_size);
+    record_offset = new_free_space - record_size;
+
+retry:
+    // If we cross the area limit, we need to invoke GC. However, we should consider all the cases
+    // where writers work in parallel, and we only want the FIRST writer to invoke GC.
+    if (new_free_space >= _size) {
+        // In the case we have crossed the limit, but the initial offset was still before the limit, this
+        // means we are the first writer (common case). Exclusively lock _lock, and invoke GC.
+        if (record_offset < _size) {
+            _lock.promote();
+            ret = garbage_collection(key, flags, buf_len_bytes, buf);
+            _lock.exclusive_unlock();
+            return ret;
+        }
+        else {
+            // In the case we have crossed the limit, and the initial offset was also after the limit,
+            // this means we are not the first writer (uncommon case). Just wait for GC to complete.
+            // then retry the operation
+            _lock.shared_unlock();
```

Comment:
I think this can be simplified to

    _lock.shared_unlock();
    rtos::Thread::wait(MEDITATE_TIME_MS);
    goto retry;

with `retry` moved above the `shared_lock()` above.

This version fails to set `record_offset` and has a no-longer-doing-anything `for (;;)`

Reply:
Right. This is the cleanest way to go.